tket/src/Circuit/ToffoliBox.cpp

```diff
@@ -0,0 +1,320 @@
+  // indices. First half of the vertices represents the row entries the second
+  // half represents the row indices
+  cube_graph_t g(1 << (n_right_columns + 1));
+  for (unsigned postfix_dec = 0; postfix_dec < (unsigned)(1 << n_right_columns);
+       postfix_dec++) {
+    std::vector<bool> postfix = dec_to_bin(postfix_dec, n_right_columns);
+    // row0 = prefix+0+postfix
+    // row1 = prefix+1+postfix
+    std::vector<bool> row0(prefix);
+    std::vector<bool> row1(prefix);
+    row0.push_back(0);
+    row0.insert(row0.end(), postfix.begin(), postfix.end());
+    row1.push_back(1);
+    row1.insert(row1.end(), postfix.begin(), postfix.end());
+    // the postfix of row0's index
+    std::vector<bool> idx0_postfix(
+        perm[row0].begin() + prefix.size() + 1, perm[row0].end());
+    // the postfix of row1's index
+    std::vector<bool> idx1_postfix(
+        perm[row1].begin() + prefix.size() + 1, perm[row1].end());
+    // row0 can stay where it is
+    boost::add_edge(
+        postfix_dec, bin_to_dec(idx0_postfix) + (1 << n_right_columns), g);
+    if (idx0_postfix != idx1_postfix) {
+      // row0 can also move to row1's index by a swap
+      boost::add_edge(
+          postfix_dec, bin_to_dec(idx1_postfix) + (1 << n_right_columns), g);
+    }
+  }
+  // find a matching
+  std::vector<perm_vert_t> match(1 << (n_right_columns + 1));
+  edmonds_maximum_cardinality_matching(g, &match[0]);
+  std::vector<std::vector<bool>> swap_pairs;
+  for (unsigned postfix_dec = 0; postfix_dec < (unsigned)(1 << n_right_columns);
+       postfix_dec++) {
+    std::vector<bool> postfix = dec_to_bin(postfix_dec, n_right_columns);
+    std::vector<bool> row0(prefix);
+    row0.push_back(0);
+    row0.insert(row0.end(), postfix.begin(), postfix.end());
+    std::vector<bool> mapped_idx = dec_to_bin(
+        match[postfix_dec] - (1 << n_right_columns), n_right_columns);
+    if (!std::equal(
+            perm[row0].begin() + prefix.size() + 1, perm[row0].end(),
+            mapped_idx.begin())) {
+      // row0 matched to row1's index
+      row0.erase(row0.begin() + col_idx);
+      swap_pairs.push_back(row0);
+    }
+  }
+  return swap_pairs;
+}
+
+static void swap_rows(
+    const std::vector<bool> &pair, unsigned col_idx, state_perm_t &perm,
+    ctrl_op_map_t &op_map, std::vector<Complex> &phases,
+    const Op_ptr &zflip_op) {
+  // swap a pair of rows along col_idx
+  // assume two rows only differ at q
+  // update op_map, phases, and permutation
+  std::vector<bool> row0(pair);
+  std::vector<bool> row1(pair);
+  row0.insert(row0.begin() + col_idx, 0);
+  row1.insert(row1.begin() + col_idx, 1);
+  std::swap(perm.at(row0), perm.at(row1));
+  // phases is a vector, convert rows to decimal to track phases
+  unsigned row0_dec = bin_to_dec(row0);
+  unsigned row1_dec = bin_to_dec(row1);
+  if (zflip_op->get_type() == OpType::Rx &&
+      equiv_val(zflip_op->get_params()[0], 1., 4)) {
```

Comment:
What are `4` and `1.` here? A quick comment why they're hardcoded as given would be great.